specs/vm/instruction_set.md

```diff
@@ -1785,3 +1786,101 @@ Panic if:
 - `$fp->$fp == 0` (if parent context is external)
 
 Set `$rA` to `$fp->$fp` (i.e. `$rA` will point to the previous call frame's contract ID).
+
+### GTF: Get transaction fields
+
+|             |                         |
+|-------------|-------------------------|
+| Description | Get transaction fields. |
+| Operation   | Varies (see below).     |
+| Syntax      | `gtf $rA, $rB, imm`     |
+| Encoding    | `0x00 rA rB i i`        |
+| Notes       |                         |
+
+Get [fields from the transaction](../protocol/tx_format.md#transaction).
+
+| name                                      | `imm`   | set `$rA` to                                     |
+|-------------------------------------------|---------|--------------------------------------------------|
+| `GTF_TYPE`                                | `0x001` | `tx.type`                                        |
+| `GTF_SCRIPT_GAS_PRICE`                    | `0x002` | `tx.gasPrice`                                    |
+| `GTF_SCRIPT_GAS_LIMIT`                    | `0x003` | `tx.gasLimit`                                    |
+| `GTF_SCRIPT_MATURITY`                     | `0x004` | `tx.maturity`                                    |
+| `GTF_SCRIPT_SCRIPT_LENGTH`                | `0x005` | `tx.scriptLength`                                |
+| `GTF_SCRIPT_SCRIPT_DATA_LENGTH`           | `0x006` | `tx.scriptDataLength`                            |
+| `GTF_SCRIPT_INPUTS_COUNT`                 | `0x007` | `tx.inputsCount`                                 |
+| `GTF_SCRIPT_OUTPUTS_COUNT`                | `0x008` | `tx.outputsCount`                                |
+| `GTF_SCRIPT_WITNESSES_COUNT`              | `0x009` | `tx.witnessesCount`                              |
+| `GTF_SCRIPT_RECEIPTS_ROOT`                | `0x00A` | Memory address of `tx.receiptsRoot`              |
+| `GTF_SCRIPT_SCRIPT`                       | `0x00B` | Memory address of `tx.script`                    |
+| `GTF_SCRIPT_SCRIPT_DATA`                  | `0x00C` | Memory address of `tx.scriptData`                |
+| `GTF_SCRIPT_INPUT_AT_INDEX`               | `0x00D` | Memory address of `tx.inputs[$rB]`               |
```

Comment:
Ditto; len?

Reply:
The length of an input or output isn't really important since it's a struct instead of a list.

Reply:
But aren't we creating a tech debt for when the user might want to iterate over all the inputs/outputs/witnesses?

Reply:
Can you clarify? If they want to iterate then they just iterate?

Reply:
Lets say the user wants to iterate all the outputs of a transaction. Currently, he can fetch the output count in a constant offset of the canonical serialization.

Since we are shielding the user from long-term breaking changes in the structure of the VM, this deployed subroutine will no longer be valid when that offset changes because it will no longer be a constant he can fetch at runtime.

So, to set the number of iterations for the outputs, that can be trivially done - and breaking-changes safe - to fetch these from this function.

Reply:
Ah I think I see the confusion. You can get the number of inputs (i.e. the length of the list of inputs) with `GTF_SCRIPT_INPUTS_COUNT`. What you can't get is the _length in bytes_ of a particular input, because as above a single input is a struct, not a list.

Reply:
Oh, overlooked that! Yes, we have these already implemented